docs/proposals/20230525-openstack-microversion-support.md

```diff
@@ -0,0 +1,133 @@
+
+Goal 1 is trickier.
+It will require CAPO to match used features with required versions.
+The outcome would affect the list of supported versions that can be used when negotiating the version with the server.
+Luckily, we only have 2 version changes so far, and one of them did not require any code changes in CAPO.
+In other words, we have Nova v2 that can be used without server tags and (2.53 or 2.60) that can be used if server tags are used.
+
+The first possible place to detect usage of a specific feature is in the webhook.
+But we do not allow API calls to OpenStack in the webhook, so this is not an option.
+Another possibility is to check at time of creation, like we do for [trunk support](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/89b6ead01a90eeca64b89cef4e0039e19c3dd93a/pkg/cloud/services/compute/instance.go#L207-L215).
+Just before getting the client and issuing the API call, we know both what we are about to do and what the input is.
+This means we can determine the version requirements for each specific call and use this information when creating the client.
+
+With this approach we get fine grained control and an implementation that can be extended with more versions as needed.
+However, it can become complex with checks sprinkled through out the code base.
+To mitigate this, the proposal is to have unified logic for determining the version requirements (per OpenStack service).
+I.e. avoid setting different versions per call just because it is possible.
+If we determine that Nova 2.53 or 2.60 is needed for some API call, for example, then use that throughout the current context.
+
+## Alternatives
+
+### Fixed version
+
+This alternative is basically to continue as before.
+The (micro)version is hard coded in CAPO so one and only one is supported and used for all API calls.
+This means that using CAPO in an environment that requires a different microversion is impossible without making a custom build.
+On the other hand, complexity is very low with this approach.
+
+### Version set by the user in clouds.yaml
+
+In this alternative the user would be able to override the version set in CAPO.
+The obvious consequence is that the user could very easily break things by using versions that are not tested or supported.
+On the other hand, it gives maximum flexibility.
+The user can use any version, but the responsibility to ensure it is working also falls on them.
+
+The `clouds.yaml` configuration file already has support for specifying the version so this should be familiar to many users.
+However, Gophercloud does not support reading all the versions from this file as of now.
+The main reason for avoiding this alternative, though, is the lack of control that CAPO has.
+With a well defined list of supported and tested versions, perhaps it could still be considered.
```

Comment:
A well defined list of supported and tested versions would be nice.
Howerver, I don't think it should be expected that our list is a complete list of all available microversions of OpenStack Cinder, Manila, Barbican, Nova and wether we tested or not.

Reply:
Very true. I'm mainly keeping this here as a reminder of where the discussion started. That is, I wanted to let the user pick the version (with all the consequences that it brings). So if we created a list of known good versions for CAPO (the list could have a single version even) then maybe we could still consider this implementation. The user could check that the version they set is supported, but in theory they could then still set an unsupported version (with the risks that it would bring).